File: 2022/Day 8/pt1.cpp
#include "helper.hpp"

int scan_tree_left(std::vector<std::vector<int>> &grid, int x, int y)
{
    int visible = 1;
    for (int tmp_x = x - 1; tmp_x >= 0; tmp_x--)
    {
        if (grid[y][tmp_x] >= grid[y][x])
            visible = 0;
    }
    return visible;
}

int scan_tree_right(std::vector<std::vector<int>> &grid, int x, int y)
{
    int visible = 1;
    for (int tmp_x = x + 1; tmp_x < grid[y].size(); tmp_x++)
    {
        if (grid[y][tmp_x] >= grid[y][x])
            visible = 0;
    }
    return visible;
}

int scan_tree_top(std::vector<std::vector<int>> &grid, int x, int y)
{
    int visible = 1;
    for (int tmp_y = y - 1; tmp_y >= 0; tmp_y--)
    {
        if (grid[tmp_y][x] >= grid[y][x])
            visible = 0;
    }
    return visible;
}

int scan_tree_bottom(std::vector<std::vector<int>> &grid, int x, int y)
{
    int visible = 1;
    for (int tmp_y = y + 1; tmp_y < grid.size(); tmp_y++)
    {
        if (grid[tmp_y][x] >= grid[y][x])
            visible = 0;
    }
    return visible;
}

int get_scenic_score(std::vector<std::vector<int>> &grid, int x, int y)
{
    int l = scan_tree_left(grid, x, y);
    int r = scan_tree_right(grid, x, y);
    int u = scan_tree_top(grid, x, y);
    int d = scan_tree_bottom(grid, x, y);
    return l || r || u || d;
}

int get_visible_trees(std::vector<std::string> lines)
{
    std::vector<std::vector<int>> grid(lines.size());

    for (int i = 0; i < lines.size(); i++)
    {
        std::string line = lines[i];
        for (char c : line)
            grid[i].push_back(c - '0');
    }

    int visible_trees = (grid.size() * 2 + grid[0].size() * 2 - 4);

    for (int y = 1; y < grid.size() - 1; y++)
    {
        for (int x = 1; x < grid[y].size() - 1; x++)
        {
            visible_trees += get_scenic_score(grid, x, y);
        }
    }

    return visible_trees;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");

    auto t1 = Helper::get_current_time();
    int visible_trees = get_visible_trees(lines);
    std::cout << "Visible Trees: " << visible_trees << std::endl;
    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/day_03/pt1.go
package main

import (
	"fmt"
	"regexp"
	"slices"
	"strconv"

	"github.com/benjababe/advent-of-code/helper"
)

type Cell struct {
	Value int64
	Id    int64
}

func isSymbol(char rune) bool {
	return !(char == '.' || ('0' <= char && char <= '9'))
}

func populateGrid(grid [][]Cell, lines []string) {
	regex := regexp.MustCompile(`(\d+)`)
	idCount := int64(1)

	for i, line := range lines {
		grid[i] = make([]Cell, len(line))
		matches := regex.FindAllStringIndex(line, -1)

		for _, match := range matches {
			numStr := line[match[0]:match[1]]
			for j := match[0]; j < match[1]; j++ {
				num, _ := strconv.ParseInt(numStr, 10, 64)
				grid[i][j] = Cell{Value: num, Id: idCount}
			}
			idCount++
		}
	}
}

func getNumsAdjacent(grid [][]Cell, i int, j int) []int64 {
	nums := []int64{}
	numIds := []int64{}

	for _i := -1; _i <= 1; _i++ {
		for _j := -1; _j <= 1; _j++ {
			di, dj := i+_i, j+_j

			if di < 0 || di >= len(grid) || dj < 0 || dj >= len(grid[di]) {
				continue
			}

			cell := grid[di][dj]
			if cell.Id > 0 && !slices.Contains(numIds, cell.Id) {
				nums = append(nums, cell.Value)
				numIds = append(numIds, cell.Id)
			}
		}
	}

	return nums
}

func searchGrid(grid [][]Cell, lines []string) int64 {
	score := int64(0)

	for i, row := range lines {
		for j, char := range row {
			if isSymbol(char) {
				for _, num := range getNumsAdjacent(grid, i, j) {
					score += num
				}
			}
		}
	}

	return score
}

func solve(lines []string) int64 {
	grid := make([][]Cell, len(lines))
	populateGrid(grid, lines)
	return searchGrid(grid, lines)
}

func main() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solve(lines)
	fmt.Printf("Output: %d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)

	helper.CopyClipboard(strconv.FormatInt(output, 10))
}
